_posts/2017-1-20-typing-the-debugger.markdown

```diff
@@ -0,0 +1,159 @@
+
+Here are some types to give you an idea of what we’re talking about here:
+
+```js
+type Breakpoint = {
+  id: string,
+  location: Location,
+  loading: boolean,
+  disabled: boolean,
+  text: string,
+  condition: ?string,
+};
+```
+
+```js
+type Location = {
+  sourceId: string,
+  line: number,
+  column?: number
+};
+```
+
+With these types we could write redux actions for **adding** and **removing** breakpoints:
+
+```js
+function addBreakpoint(location: Location) {
+  // ...
+}
+
+function removeBreakpoint(breakpoint: Breakpoint) {
+  // ...
+}
+```
+
+There are a couple advantages to this style:
+1. the code is slightly clearer to read
+2. it’s clear that breakpoint param can never be a string or undefined
+3. anyone can look up the `Breakpoint` type and see what data it has. I do this *all* the time now!
+4. we get a warning if we accidentally call `addBreakpoint` without passing a location in or we try to pass too many arguments in.
+
+
+### Typing the Debugger - story time
+
+In the summer we set what we thought was a modest [goal](.) of adding types in all of the major components of our app: actions, reducers, UI components, utils, and browser client. We just reached it, which is a sign that typing a sufficiently large application does take time and you should expect to do it incrementally.
+
+Here’s the story of how we added types. There are some interesting lessons here and there are certainly a **lot** of people to thank.
+
+**Step 1:  go to the center**
+
+The core of the Debugger is the client that speaks with the browser. The Debugger started off with two simple modules `firefox.js` and `chrome.js` which were responsible for making a websocket connection, sending API commands, and receiving notifications. One *great* decision we made early on was that chrome and firefox would share the same API interface, so basically there would be only one way to set a breakpoint. James’ started adding types here [1](https://github.com/devtools-html/debugger.html/pull/213/files),   [2](https://github.com/devtools-html/debugger.html/pull/222/files). The rationale was that if we got reliable data back from the client the rest of the Debugger would be a whole lot saner. In fact, if Chrome and Firefox shared the same API, then as long as we built the Chrome client to spec nothing else should change. This is still true today.
+
+**Step 2: It’s about the data stupid**
+
+We chose to use  **TComb** for typing the Application in May. We chose it because we wanted a runtime type checker. At the time, we were not sure if would have a build step and using a runtime type checker meant that we would not be changing the syntax of the code. I also believed at the time that a runtime type checker would catch more issues. I was significantly underestimated how awesome **Flow** is.
```

Comment:
There's also a babel plugin for converting flow types to tcomb btw: https://github.com/gcanti/babel-plugin-tcomb, there's also flow-runtime: https://codemix.github.io/flow-runtime/#/